src/solve.jl

```diff
@@ -17,71 +17,493 @@ function DiffEqBase.__solve(prob::BVProblem, alg::Shooting; kwargs...)
         ReturnCode.Failure)
 end
 
-function DiffEqBase.__solve(prob::BVProblem, alg::Union{GeneralMIRK, MIRK}; dt = 0.0,
-    kwargs...)
+function DiffEqBase.__solve(prob::BVProblem, alg::Union{GeneralMIRK, MIRK}; dt = 0.0, abstol = 1e-6,
+                            kwargs...)
     dt ≤ 0 && throw(ArgumentError("dt must be positive"))
     n = Int(cld((prob.tspan[2] - prob.tspan[1]), dt))
-    x = collect(range(prob.tspan[1], stop = prob.tspan[2], length = n + 1))
-    S = BVPSystem(prob, x, alg_order(alg))
-
-    tableau = constructMIRK(S)
-    cache = alg_cache(alg, S)
-    # Upper-level iteration
-    vec_y = Array{eltype(first(S.y))}(undef, S.M * S.N)              # Vector
-    function reorder!(resid)
-        # reorder the Jacobian matrix such that it is banded
-        tmp_last = resid[end]
-        for i in (length(resid) - 1):-1:1
-            resid[i + 1] = resid[i]
-        end
-        resid[1], resid[end] = resid[end], tmp_last
-        return nothing
+    mesh = collect(range(prob.tspan[1], stop = prob.tspan[2], length = n + 1))
+    defect_threshold = 0.1
+    info = 0
+    defect_norm = 10
+    MxNsub = 3000
+    while info == 0 && defect_norm > tol
+        S = BVPSystem(prob, mesh, alg_order(alg))
+        len = S.M
+        tableau = constructMIRK(S)
+        cache = alg_cache(alg, S)
+        # Upper-level iteration
+        vec_y = Array{eltype(first(S.y))}(undef, S.M * S.N)              # Vector
+        function reorder!(resid)
+            # reorder the Jacobian matrix such that it is banded
+            tmp_last = resid[end]
+            for i in (length(resid) - 1):-1:1
+                resid[i + 1] = resid[i]
+            end
+            resid[1], resid[end] = resid[end], tmp_last
+            return nothing
+        end
+        function loss!(resid, u0, p)
+            nest_vector!(S.y, u0)
+            @set! S.p = p
+            Φ!(S, tableau, cache)
+            if isa(prob.problem_type, TwoPointBVProblem)
+                eval_bc_residual!(S)
+            else
+                general_eval_bc_residual!(S)
+            end
+            flatten_vector!(resid, S.residual)
+            reorder!(resid)
+            return nothing
+        end
+    
+        jac_wrapper = BVPJacobianWrapper(loss!)
+
+        flatten_vector!(vec_y, S.y)
+        nlprob = _construct_nonlinear_problem_with_jacobian(jac_wrapper, S, vec_y, prob.p)
+        opt = solve(nlprob, alg.nlsolve; kwargs...)
+        nest_vector!(S.y, opt.u)
+
+        k_discrete = copy(cache.k_discrete)
+
+        if opt[2] == ReturnCode.Success
+            info = 0
+        elseif opt[2] == ReturnCode.Failure
+            info = 1
+        end
+
+        # Change the original solution Vector{Vector{Float64}} to a matrix
+        # has size of (number of subintervals, length of equation)
+        global Y = transpose(reduce(hcat, S.y))
```

Comment:
why is this global?

Reply:
The discrete solution ```Y``` is a local variable in the loop of iteration, so I set ```Y``` as global so we can finally build a solution based on ```Y```.